man/zshlogger.1
.TH zshlogger "1" "DATE" "MegaCoder tools" "User Commands"
.SH NAME
zshlogger \- pass commands to a \fBzsh\fP(\fI1\fP) and control its output
.SH SYNPOSIS
.B zshlogger
[\fIoptions\fP]
.SH DESCRIPTION
.PP
\fBzshlogger\fP(\fI1\fP) passes commands to a \fBzsh\fP(\fI1\fP) instance
and records any output into a file.
This is quite similar to \fBscript\fP(\fI1\fP) with an important difference:
the output of some commands is hidden and phony results can be placed into
the output.
.PP
When producing camera-ready artwork for technical classes, sometimes the
normal output of a command must be massaged.  Any irrelevant preparator commands
should be executed, but they shouldn't produce any distracting output.
.PP
To this end, \fBzshlogger\fP(\fI1\fP) recognizes a minimal scripting
language.
Anything not recognized as a scripting command is passed to a slave
\fBzsh\fP(\fI1\fP) instance; any output the command produces is copied
to a log file.
.TP
\fB\-D\fP
increase debug output verbosity; useful only to the author and maybe not even
him.
.TP
\fB\-a\fP
prefix each executed command with its absolute path.
Anything run by \fIroot\fP should probably display the absolute path to
the commands, to encourage best system administrator practice.
.TP
\fB\-c\fP \fIcmd\fP
execute the argument \fIcmd\fP only and then exit.
This is similar behavior to all shells, such as \fBbash\fP(\fI1\fP) or
\fBzsh\fP(\fI\fP).
.TP
\fB\-f\fP \fIscript\fP
read \fBzshlogger\fP(\fI1\fP) commands from \fIscript\fP instead of from
\fIstdin\fP.
.TP
\fB\-h\fP
produce a succinct usage summary.
.TP
\fB\-o\fP \fIofile\fP
write all output to \fIofile\fP instead of to \fIstdout\fP.
.TP
\fB\-p\fP \fIprompt\fP
have the slave \fBzsh\fP(\fI1\fP) use \fIprompt\fP as the shell prompt instead
of whatever it would normally use.
Usefull for having commands appear to be run by \fIroot\fP even if the script
is run by another user.
.TP
\fB\-s\fP \fIshell\fP
use \fIshell\fP as the slave shell instead of the default \fB/bin/zsh\fP;
just in case your shell is somewhere else.
A non-zsh shell would probably not work.
.SH "SCRIPTING LANGUAGE"
.PP
The scripting language is really small.
An octothorpe (\fB#\fI) in column 1 begins a scripting command:
.TP
\fB#ignore\fP
ignore this line completely.
.TP
\fB#hide\fP \fIcommand...\fP
Perform \fIcommand\fP but show neither the command nor the output.
Use this to silently prepare for a real command; create temporary files;
download stuff; whatever you need to hide.
.TP
\fB#note\fP \fITitle\fP
insert editorial comment in the output.
.TP
\fB#snip\fP \fIcommand...\fP
show and run the command but do \fInot\fP display any results.
.TP
\fBquiet\fP \fIcommand\fP
run the \fIcommand\fP but do not show the command.
Show the results.
.TP
\fB#step\fP \fIdescription\fP
insert editorial command into the output.
These comments are sequentially numbered to serve as section titles.
.SH AUTHOR
Written by Tommy Reynolds <Tommy.Reynolds@MegaCoder.com>
.SH REPORTING BUGS
Bugs?  What are you thinking?
.SH COPYRIGHT
Copyright \(co 2013 Tommy Reynolds <Tommy.Reynolds@MegaCoder.com>
.br
International copyright secured.  All rights reserved.
.br
This software is licensed under the GPLv2, exactly.
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
All warranties are declaimed and denied, including any implied warranty of merchantibility.
This software has no claim of fitness for any purpose


ChangeLog
2010-03-09 18:34  reynolds

	* cvs2cl: Imported

2010-03-09 18:33  reynolds

	* Makefile: RPM now builds

2010-03-09 18:32  reynolds

	* Makefile.am, autogen.sh: Checkpoint

2010-03-09 18:26  reynolds

	* AUTHORS, ChangeLog, NEWS, README, autogen.sh, Makefile,
	Makefile.am, zshlogger.spec: Checkpoint

2010-03-09 18:23  reynolds

	* autogen.sh, configure.in: Checkpoint

2010-03-09 18:10  reynolds

	* Makefile.am, autogen.sh, configure.in, zshlogger.spec.in:
	Checkpoint

2010-03-09 18:06  reynolds

	* configure.in, zshlogger.spec.in: Checkpoint

2010-03-09 17:41  reynolds

	* Makefile, zshlogger.spec: Checkpoint

2010-03-09 17:38  reynolds

	* Makefile, zshlogger.c: Don't run past end of single-token
	commands.

2010-03-05 21:46  reynolds

	* Makefile: Modernized

2009-11-23 20:05  reynolds

	* Makefile: Use funky compile flags

2009-07-02 22:04  reynolds

	* Makefile, zshlogger.c: Use F11's POSIX-compatible symbols.

2009-02-10 14:47  treynold

	* zshlogger.c: Added "#find" splat to just locate commands if "-n"
	isn't used.

2008-12-30 10:57  treynold

	* zshlogger.zsh, zshlogger2zsh.zsh: Decruft

2008-09-04 06:52  reynolds

	* zshlogger.c: 1)  Added '-h' and '-H' homer switches.	2)  Added
	'-c cmd' switch, too.

2008-08-30 11:39  reynolds

	* zshlogger.c: Added '-a' switch to turn off displaying absolute
	paths.

2008-08-30 11:37  reynolds

	* Makefile, zshlogger.c: If first command line token is an
	executable on the ${PATH}, show its absolute path.

2008-08-30 10:39  reynolds

	* zshlogger.c: 1)  Added '-p prompt' switch.  2)  Dropped leading
	\n from the EOF message.

2008-08-30 10:29  reynolds

	* Makefile: Discarded forked zsh scripts.

2008-08-30 10:27  reynolds

	* .cvsignore, Makefile, test, zshlogger.c: Use ZSH to build
	multi-step scripts.

2005-12-03 12:35  reynolds

	* test, zshlogger.c: Changed mind on the "#fixme" keyword: it's now
	"#ignore".

2005-12-03 12:34  reynolds

	* Makefile, test, zshlogger.c: Made keyword recognition
	case-insentitive.  Added "#fixme" keyword to completely ignore a
	line.

2005-10-05 13:00  reynolds

	* zshlogger2zsh.zsh: Don't put the shell prompt in the output, so
	we could actually feed the script to a ZSH process.

2005-10-05 12:57  reynolds

	* Makefile, test: 1) Can't strip a shell script, Dolt! 2) Need to
	escape single quotes in #note lines.

2005-10-05 12:54  reynolds

	* unscript.zsh, zshlogger2zsh.zsh: Renamed the freakin' script.

2005-10-05 12:53  reynolds

	* test, unscript.zsh: Convert a zshlogger script into an ordinary
	ZSH script.

2005-09-20 18:26  reynolds

	* Makefile, test, zshlogger.c: Since readline() is dropping TAB
	characters from the input, use fgets() iff input is from a file.

2005-08-18 14:50  reynolds

	* zshlogger.c: Added "#note" for adding editorial comments.

2005-08-18 14:05  reynolds

	* Makefile, test, zshlogger.c: Doh, check the _correct_ line
	length...

2005-08-18 13:43  reynolds

	* .cvsignore: Ignorables

2005-08-18 13:43  reynolds

	* test: Need a line continuation test.

2005-08-18 13:37  reynolds

	* Makefile, zshlogger.c: 1) Use readline() to read from stdin 2)
	Allow final \\ for command line continuation

2005-08-16 14:38  reynolds

	* Makefile: Install into "/opt/bin" be default.

2005-01-17 08:22  reynolds

	* test, zshlogger.c: Added "#step" directive to provide in-context
	labels.

2005-01-03 10:01  reynolds

	* zshlogger.c: Refined script parsing so we can use whitespaces
	after #-tokens.

2004-12-12 12:26  reynolds

	* Makefile: Put into ~/bin since that doesn't get blasted when I
	re-install the OS.

2004-09-26 15:56  reynolds

	* zshlogger.c: Pretty printed

2004-09-26 15:56  reynolds

	* zshlogger.c: Added '#command' descriptions.

2004-09-26 14:36  reynolds

	* test, zshlogger.c: Don't try to escape the line; supply it to cat
	as a literal HERE document!

2004-09-22 11:59  reynolds

	* Makefile, test, zshlogger.c: Added "#quiet" directive to omit
	copying the command itself.

2004-09-22 11:23  reynolds

	* zshlogger.c: Added "usage()" routine.

2004-09-10 13:51  reynolds

	* zshlogger.c: Removed extra space after prompt.

2004-09-09 20:42  reynolds

	* Makefile, zshlogger.c: Removed extraneous redirections in the
	script.

2004-09-09 20:33  reynolds

	* zshlogger.c: Build up a shell script and then run that, recording
	the results.

2004-09-09 19:09  reynolds

	* zshlogger.c: Checkpoint.

2004-09-09 16:52  reynolds

	* Makefile, test, zshlogger.c: Switched to C program so I can
	control it better.

2004-09-09 12:38  reynolds

	* zshlogger, zshlogger.zsh: Renamed the file.

2004-09-09 12:37  reynolds

	* zshlogger: Checkpoint

2004-09-09 10:53  reynolds

	* test: Foolish test case.

2004-09-09 10:52  reynolds

	* zshlogger: Omit saving stdout from the 1 command following the
	line '#omit'.

2004-09-09 10:40  reynolds

	* zshlogger: Added the '-t term' switch so I can use simple
	underlining and such if I want.

2004-09-09 10:33  reynolds

	* Makefile, README.txt, zshlogger: Runs shell commands and captures
	their output.  Optionally converts it to XML (needs txt2xml
	helper).

2004-09-09 10:33  reynolds

	* Makefile, README.txt, zshlogger: Initial revision



src/zshlogger.c
#include <sys/types.h>
#include <sys/wait.h>
#include <unistd.h>
#include <limits.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <string.h>
#include <assert.h>
#include <errno.h>
#include <ctype.h>
#include <regex.h>
#include <readline/readline.h>
#include <readline/history.h>

typedef	struct	prompt_s	{
	int		doPrompt;
	char		thePrompt[ 48 ];
	char		extraPrompt[ 48 ];
} prompt_t;

typedef	enum	homer_e		{
	HOMER_none	= 0,
	HOMER_tilde,
	HOMER_macro
} homer_t;

static unsigned		nonfatal;
static char *		me;
static unsigned		debugLevel;
static char *		ifile;
static char *		ofile;
static char *		prompt;
static char *		shell;
static int		verbose;

static	prompt_t	promptInfo;

static FILE *		scriptFyle;
static char		scriptFn[ 256 ];
static char *		paths[ (BUFSIZ/2) + 1 ];
static unsigned		Npaths;
static unsigned		showAbsolutePath = 1;
static unsigned		runCommand = 1;
static regex_t		pattern;
static homer_t		homer = HOMER_tilde;
static char *		onlyCommand;

static void
vsay(
	int		e,
	char *		fmt,
	va_list		ap
)
{
	fprintf( stderr, "%s: ", me );
	vfprintf( stderr, fmt, ap );
	if( e )	{
		fprintf( stderr, "; errno=%d (%s)", e, strerror( e ) );
	}
	fprintf( stderr, ".\n" );
}

static int inline
doDebug(
	int		level
)
{
	return( level <= debugLevel );
}

static void
debug(
	int		level,
	char *		fmt,
	...
)
{
	if( doDebug( level ) )	{
		va_list	ap;

		va_start( ap, fmt );
		vsay( 0, fmt, ap );
		va_end( ap );
	}
}

static void
panic(
	int		e,
	char *		fmt,
	...
)
{
	va_list		ap;

	va_start( ap, fmt );
	vsay( e, fmt, ap );
	va_end( ap );
	exit( 1 );
	/*NOTREACHED*/
}

static void
usage(
	char *		fmt,
	...
)
{
	if( fmt )	{
		va_list	ap;

		fprintf( stderr, "%s: ", me );
		va_start( ap, fmt );
		vfprintf( stderr, fmt, ap );
		va_end( ap );
		fprintf( stderr, "\n" );
	}
	fprintf(
		stderr,
		"usage: %s"
		" [-D]"
		" [-a]"
		" [-c cmd]"
		" [-f script]"
		" [-h]"
		" [-o ofile]"
		" [-p prompt]"
		" [-s shell]"
		" [-v]"
		"\n",
		me
	);
	fprintf(
		stderr,
		"\t#ignore - ignore this line completely.\n"
		"\t#hide   - don't show command or results.\n"
		"\t#note   - insert editorial comment.\n"
		"\t#snip   - show and run command, elide results.\n"
		"\t#quiet  - don't show command, display results.\n"
	);
	exit( 1 );
	/*NOTREACHED*/
}

static	void
getPaths(
	void
)
{
	char *		s;
	char *		bp;

	s = strdup( getenv( "PATH" ) );
	for(
		bp = s;
		(paths[ Npaths ] = strtok( bp, ":" )) != NULL;
		bp = NULL
	)	{
		++Npaths;
	}
}

static	char *
findPath(
	char * const	token
)
{
	static	char	path[ (PATH_MAX +1) * 2 ];
	char *		retval;

	retval = token;
	if( showAbsolutePath )	{
		char * *	pp;

		for( pp = paths; *pp; ++pp )	{
			sprintf( path, "%s/%s", *pp, token );
			if( access( path, X_OK ) == 0 )	{
				retval = path;
				break;
			}
		}
	}
	return( retval );
}

static void
getProcessName(
	char *		s
)
{
	char *		bp;

	if( (bp = strrchr( s, '/' )) != NULL )	{
		s = bp + 1;
	}
	me = strdup( s );
	assert( me );
	if( (bp = strchr( me, '.' )) != NULL )	{
		*bp = '\0';
	}
}

static unsigned
processArgs(
	int		argc,
	char * *	argv
)
{
	int		c;
	unsigned	results;

	results = 0;
	while( (c = getopt( argc, argv, "DHac:f:hno:p:s:v" )) != EOF )	{
		switch( c )	{
		default:
			fprintf( stderr, "%s: no -%c yet!\n", me, c );
			/*FALLTHRU*/
		case '?':
			++results;
			break;
		case 'D':
			++debugLevel;
			break;
		case 'H':
			homer = HOMER_macro;
			break;
		case 'a':
			showAbsolutePath = 0;
			break;
		case 'c':
			onlyCommand = optarg;
			break;
		case 'f':
			ifile = optarg;
			break;
		case 'h':
			homer = HOMER_none;
			break;
		case 'n':
			runCommand = 0;
			break;
		case 'o':
			ofile = optarg;
			break;
		case 'p':
			prompt = strdup( optarg );
			break;
		case 's':
			shell = optarg;
			break;
		case 'v':
			++verbose;
			break;
		}
	}
	return( results );
}

static void
setPrompt(
	void
)
{
	uid_t const	uid = geteuid();

	prompt = uid ? "$" : "#";
}

static void
writeScript(
	char *		fmt,
	...
)
{
	char	line[ BUFSIZ ];
	va_list		ap;

	va_start( ap, fmt );
	vsnprintf( line, sizeof( line ), fmt, ap );
	if( verbose )	{
		static int	lineno = 1;
		fprintf( stderr, "line %d\t%s\n", lineno++, line );
	}
	fprintf( scriptFyle, "%s\n", line );
	va_end( ap );
}

static int
isToken(
	char const * const	buf,
	char const * const	s
)
{
	int const	len = strlen( s );
	int		results;

	results = 0;
	do	{
		if( strncasecmp( buf, s, len ) )	{
			break;
		}
		if( !isspace( buf[len] ) )	{
			break;
		}
		results = 1;
	} while( 0 );
	return( results );
}

static char *
skipWhitespace(
	char *		s
)
{
	while( *s && isspace( *s ) )	{
		++s;
	}
	return( s );
}

static unsigned
doCommand(
	char *		line
)
{
	int		hideCommand;
	int		snipResults;
	int		beQuiet;
	int		reallyRun;
	char *          bp;
	char *		verb;
	char *		rest;
	size_t		len;

	hideCommand = 0;
	snipResults = 0;
	beQuiet = 0;
	reallyRun = runCommand;
	do	{
		/* Trim trailing whitespace				 */
		line = skipWhitespace( line );
		for( bp = line + strlen( line ) - 1; bp > line; )	{
			if( !isspace( *bp ) )	{
				break;
			}
			*bp-- = '\0';
		}
		debug( 2, "trimmed=|%s|", line );
                /*
                 *--------------------------------------------------------
                 * Process flags, if any.
                 *--------------------------------------------------------
                 * #ignore - ignore this line completely
		 *
                 * #hide - run command, hide command and results.
                 *
                 * #snip - display and run command, elide results.
                 *
                 * #quiet - do not announce but run anyway.
		 *
		 * #step  - identify interesting places in logged output
		 *
		 * #note  - insert parenthetical aside into output
		 *
		 * #find  - Locate and show command but do not run
                 *--------------------------------------------------------
                 */
		if( isToken( line, "#ignore" ) )	{
			goto Exit;
		} else if( isToken( line, "#hide" ) )	{
			line += 5;
			hideCommand = 1;
			beQuiet = 1;
			snipResults = 1;
		} else if( isToken( line, "#snip" ) )	{
			line += 5;
			snipResults = 1;
		} else if( isToken( line, "#quiet" ) )	{
			line += 6;
			hideCommand = 1;
		} else if( isToken( line, "#step" ) )	{
			static unsigned	step_no = 1;
			static char	new[ BUFSIZ ];
			char		label[ 32 ];

			line = skipWhitespace( line+5 );
			snprintf( label, sizeof( label ), "Step %d:", step_no );
			snprintf( new, sizeof( new ),
				"echo -e '%s%-8s %s\n'",
				(step_no == 1) ? "" : "\n",
				label,
				line
			);
			++step_no;
			line = new;
			hideCommand = 1;
		} else if( isToken( line, "#note" ) )	{
			static char	new[ BUFSIZ ];

			line = skipWhitespace( line + 5 );
			snprintf( new, sizeof( new ),
				"echo -e '>> %s <<\n'",
				line
			);
			line = new;
			hideCommand = 1;
		} else if( isToken( line, "#find" ) )	{
			line += 5;
			hideCommand = 0;
			reallyRun = 0;
		} else if( line[0] == '#' )	{
			/* Any other splat is not ours			 */
			break;
		}
		line = skipWhitespace( line );
	} while( isspace( line[0] ) || (line[0] == '#') );
	len = strlen( line );
	verb = strtok( line, " \t" );
	if( strlen( verb ) < len )	{
		/* More stuff at end of command, step over verb		*/
		rest = verb + strlen( verb ) + 1;
	} else	{
		/* Command is one token					*/
		rest = "";
	}
	verb = findPath( verb );
        if( !hideCommand )      {
		writeScript( "cat <<-'EOF'" );
		writeScript( "%s %s %s", prompt, verb, rest );
		writeScript( "EOF" );
        }
	if( reallyRun )	{
		writeScript(
			"%s %s %s",
			verb,
			rest,
			(snipResults ? ">/dev/null 2>&1" : "")
		);
		if( snipResults && !beQuiet )	{
			writeScript( "echo '<output snipped>'" );
		}
	}
Exit:
	return( 0 );
}

static void
getShell(
	void
)
{
	shell = getenv( "SHELL" );
	if( !shell )	{
		shell = "/bin/sh";
	}
}

static void
removeScriptFile(
	void
)
{
	if( scriptFn[0] )	{
		if( debugLevel > 1 )	{
			fprintf(
				stderr,
				"%s: leaving '%s' for your inspection.\n",
				me,
				scriptFn
			);
		} else	{
			unlink( scriptFn );
		}
	}
}

static void
openScriptFile(
	void
)
{
	int		fd;
	char *		tmpdir;

	tmpdir = getenv( "TMP" );
	if( !tmpdir )	{
		tmpdir = getenv( "TMPDIR" );
		if( !tmpdir )	{
			tmpdir = "/tmp";
		}
	}
	snprintf(
		scriptFn,
		sizeof( scriptFn ),
		"%s/scr.%d.tmp.XXXXXX",
		tmpdir,
		getpid()
	);
	debug( 1, "template=%s", scriptFn );
	fd = mkstemp( scriptFn );
	if( fd < 0 )	{
		panic( errno, "cannot create temp file '%s'", scriptFn );
	}
	debug( 1, "scriptFn=%s", scriptFn );
	scriptFyle = fdopen( fd, "r+t" );
	if( !scriptFyle )	{
		panic( errno, "fdopen failed" );
	}
	if( atexit( removeScriptFile ) )	{
		panic( errno, "could not register removeScriptFile" );
	}
}

static void
initPrompt(
	prompt_t *	tp
)
{
	tp->doPrompt = isatty( 0 );
	if( tp->doPrompt )	{
		snprintf( tp->thePrompt, sizeof( tp->thePrompt ), "%s> ", me );
		snprintf( tp->extraPrompt, sizeof( tp->extraPrompt ), "%s(cont)> ", me );
	}
}

static char *
getCommand(
	prompt_t *	tp
)
{
	int		len;
	char *		theLine;
	char *		part;

	len = 0;
	theLine = NULL;
	do	{
		int	n;

		/* Get a command segment				 */
		if( tp->doPrompt )	{
			/* Input from a terminal			 */
			part = readline(
				len ? tp->extraPrompt : tp->thePrompt
			);
		} else	{
			char *	buf;

			buf = part = malloc( BUFSIZ );
			if( !buf )	{
				panic( errno, "out of memory" );
				/*NOTREACHED*/
			}
			part = fgets( buf, BUFSIZ, stdin );
			if( !part )	{
				/* EOF, don't leak buffer		 */
				free( buf );
			} else	{
				/* Drop line terminator			 */
				part[ strlen( part ) - 1 ] = '\0';
			}
		}
		if( !part )	{
			break;
		}
		debug( 2, "read part |%s|", part );
		n = strlen( part );
		/* Append part to end of the whole line			 */
		theLine = realloc( theLine, len + n + 1 );
		if( !theLine )	{
			free( part );
			break;
		}
		strncpy( theLine + len, part, n );
		len += n;
		theLine[ len ] = '\0';
		free( part );
		debug( 2, "input so far |%s|", theLine );
		/* Break out if line is not continued			 */
		if( theLine[ len - 1 ] != '\\' )	{
			break;
		}
		theLine[ len - 1 ] = ' ';
	} while( theLine );
	return( theLine );
}

static void
freeCommand(
	char *		cmd
)
{
	free( cmd );
}

static	void
setupHomer(
	void
)
{
	char * const	home = getenv( "HOME" );
	char		buf[ BUFSIZ + 1 ];

	/* Divide into three parts: leadin, ${HOME}, trailer		 */
	sprintf( buf, "^\\(.*\\)\\(%s\\)\\(.*\\)$", home );
	/* Compile into the pattern buffer				 */
	if( regcomp( &pattern, buf, REG_EXTENDED ) )	{
		perror( buf );
		exit( 1 );
	}
	debug( 1, "homer-pattern=|%s|", buf );
}

static	void
printPart(
	char * const		s,
	regmatch_t * const	rm
)
{
	printf( "%.*s", rm->rm_eo - rm->rm_so, s + rm->rm_so );
}

int
main(
	int		argc,
	char * *	argv
)
{
	char		line[ BUFSIZ ];
	FILE *		cmd;

	getProcessName( argv[0] );
	getPaths();
	getShell();
	setPrompt();
	nonfatal += processArgs( argc, argv );
	if( nonfatal )	{
		usage( "illegal argument(s)" );
		/*NOTREACHED*/
	}
	setupHomer();
	if( ofile )	{
		unlink( ofile );
		if( freopen( ofile, "wt", stdout ) != stdout )	{
			panic( errno, "cannot redirect output to '%s'",
				ofile );
		}
	}
	openScriptFile();
	if( onlyCommand )	{
		nonfatal += doCommand( onlyCommand );
	} else if( optind < argc )	{
		/* Take command from the command line		 */
		int	i;
		char *	bp;
		char *	lbp;

		bp = line;
		lbp = bp + sizeof( line );
		for( i = optind; i < argc; ++i )	{
			bp += snprintf(
				bp,
				lbp - bp,
				"%s%s",
				((i == optind) ? "" : " "),
				argv[ i ]
			);
		}
		nonfatal += doCommand( line );
	} else	{
		/* Read commands from stdin, redirecting if needed	 */
		if( ifile &&
		(freopen( ifile, "rt", stdin ) != stdin ) )	{
			panic( errno, "cannot redirect input to '%s'",
					ifile );
			/*NOTREACHED*/
		}
		initPrompt( &promptInfo );
		while( !feof( stdin ) )	{
			char *		line;

			/* Read input line			 */
			line = getCommand( &promptInfo );
			if( !line )	{
				if( promptInfo.doPrompt )	{
					fprintf( stderr, " EOF\n" );
				}
				break;
			}
			nonfatal += doCommand( line );
			freeCommand( line );
		}
	}
	fclose( scriptFyle );
	snprintf(
		line,
		sizeof( line ),
		"%s %s 2>&1",
		shell,
		scriptFn
	);
	cmd = popen( line, "r" );
	if( !cmd )	{
		panic( errno, "could not open pipe '%s'", line );
	}
	while( fgets( line, sizeof( line ), cmd ) )	{
#		define		PATTERNS	4
		regmatch_t	parts[PATTERNS];

		if( homer == HOMER_none )	{
			fputs( line, stdout );
		}  else	{
			if( !regexec( &pattern, line, PATTERNS, parts, 0 ) ) {
				regmatch_t * const	leadin = parts + 1;
				regmatch_t * const	trailer = parts + 3;

				printPart( line, leadin );
				printf(
					(homer == HOMER_macro) ? "${HOME}" :
						"~"
				);
				printPart( line, trailer );
				printf( "\n" );
			} else	{
				fputs( line, stdout );
			}
		}
	}
	pclose( cmd );
	return( nonfatal ? 1 : 0 );
}
